tools/chpldef/Message.h

```diff
@@ -0,0 +1,396 @@
+
+  /** If this message failed, return a possible note explaining why. */
+  const std::string& note() const { return note_; }
+
+  /** Get the numeric value of an error code. */
+  static inline int64_t errorToInt(Error e) {
+    return static_cast<int64_t>(e);
+  }
+
+  /** Get the numeric value of an error code. */
+  inline int64_t errorToInt() const { return errorToInt(error_); }
+
+  /** Return the current status of this message. */
+  inline Message::Status status() const { return status_; }
+
+  /** Returns 'true' if this message is marked 'COMPLETED'. */
+  inline bool isCompleted() const { return status_ == Message::COMPLETED; }
+
+  /** If 'true', this message is a response. */
+  inline bool isResponse() const { return tag_ == Message::RESPONSE; }
+
+  /** If 'true', the server is sending this message to the client. */
+  bool isOutbound() const;
+
+  /** If 'true', then this message does not need a response. */
+  bool isNotification() const;
+
+  /** Returns the expected JSON-RPC method name for this message. */
+  const char* jsonRpcMethodName() const;
+
+  /** Returns the message ID. */
+  inline const JsonValue& id() const { return id_; }
+
+  /** Returns the message ID as a string. */
+  std::string idToString() const;
+
+  /** Convert a given message tag to string. */
+  static const char* tagToJsonRpcMethodName(Tag tag);
+
+  /** Convert from JSON-RPC method name to message tag. */
+  static Tag jsonRpcMethodNameToTag(const char* str);
+
+  /** Convert from JSON-RPC method name to message tag. */
+  static Tag jsonRpcMethodNameToTag(std::string str);
+
+  /** Dynamic cast to a response. */
+  const Response* toResponse() const;
+
+  /** Dynamic cast to a response, mutable. */
+  Response* toResponse();
+
+  /** Dynamic cast to other message subclasses, const and non-const. */
+  #define CHPLDEF_MESSAGE(name__, x1__, x2__, x3__) \
+  const class name__* to##name__() const; \
+  class name__* to##name__();
+  #include "./message-macro-list.h"
+  #undef CHPLDEF_MESSAGE
+
+  /** Base visitor over messages, does nothing and returns default value. */
+  template <typename T>
+  class Visitor {
+  private:
+    Server* ctx_;
+
+  public:
+    Visitor(Server* ctx) : ctx_(ctx) {} 
+
+    virtual T visit(const class Response* rsp) {
+      T ret;
+      return ret;
+    }
+
+    /** Default visitor body is trivial enough that we define it here. */
+    #define CHPLDEF_MESSAGE(name__, x1__, x2__, x3__) \
+    virtual T visit(const class name__* req) { \
+      T ret; \
+      return ret; \
+    }
+    #include "./message-macro-list.h"
+    #undef CHPLDEF_MESSAGE
+
+    /** Invoke the visitor on a given message, defined below. */
+    T dispatch(const Message* req);
+
+    /** Get the server context. */
+    Server* ctx() const;
+  };
+
+  /** Dispatch a visitor using a message as the receiver. */
+  template <typename T, typename U>
+  inline T accept(Message::Visitor<U>& v) { return v.dispatch(*this); }
```

Comment:
Why is the T different from the U here? Doesn't Visitor<U>::dispatch return U?

Reply:
Good point.